cachito/workers/pkg_managers/yarn.py

```diff
@@ -55,3 +64,289 @@ def get_yarn_proxy_repo_username(request_id):
+        {<dependency identifier>: <dependency info>}
+    :return: copy of yarn.lock data with replacements applied
+    """
+    yarn_lock_new = copy.deepcopy(yarn_lock)
+    yarn_lock_new.update(copy.deepcopy(nexus_replacements))
+    return yarn_lock_new
+
+
+def _sanity_check(app_source_path):
+    """
+    Sanity-check the application directory.
+
+    - package.json must be present
+    - yarn.lock must be present
+    - package-lock.json must not be present
+    # TODO: npm-shrinkwrap.json?
+
+    :param Path app_source_path: the full path to the application source code
+    :raise ValidationError: if any of the above-mentioned conditions are not met
+    """
+    package_json = app_source_path / "package.json"
+    yarn_lock = app_source_path / "yarn.lock"
+    package_lock_json = app_source_path / "package-lock.json"
+
+    if not package_json.is_file():
+        raise ValidationError("The package.json file must be present for the yarn package manager")
+
+    if not yarn_lock.is_file():
+        raise ValidationError("The yarn.lock file must be present for the yarn package manager")
+
+    if package_lock_json.is_file():
+        raise ValidationError(
+            "Found a package-lock.json file while processing a yarn package. "
+            "Aborting to avoid ambiguity."
+        )
+
+
+def resolve_yarn(app_source_path, request, skip_deps=None):
+    """
+    Resolve and fetch npm dependencies for the given app source archive.
+
+    :param (str | Path) app_source_path: the full path to the application source code
+    :param dict request: the Cachito request this is for
+    :param set skip_deps: a set of dependency identifiers to not download because they've already
+        been downloaded for this request
+    :return: a dictionary that has the following keys:
+        ``deps`` which is the list of dependencies,
+        ``downloaded_deps`` which is a set of the dependency identifiers of the dependencies that
+        were downloaded as part of this function's execution,
+        ``lock_file`` which is the lock file if it was modified,
+        ``package`` which is the dictionary describing the main package, and
+        ``package.json`` which is the package.json file if it was modified.
+    :rtype: dict
+    :raises CachitoError: if fetching the dependencies fails or required files are missing
+    """
+    app_source_path = Path(app_source_path)
+    _sanity_check(app_source_path)
+
+    package_json_path = app_source_path / "package.json"
+    yarn_lock_path = app_source_path / "yarn.lock"
+    package_and_deps_info = _get_package_and_deps(package_json_path, yarn_lock_path)
+
+    # By downloading the dependencies, it stores the tarballs in the bundle and also stages the
+    # content in the npm repository for the request
+    proxy_repo_url = get_yarn_proxy_repo_url(request["id"])
+    package_and_deps_info["downloaded_deps"] = download_dependencies(
```

Comment:
I just realized that `download_dependencies` always download to `deps/npm`, we probably want to change that right?

Reply:
Yes, it would be nice to pass a parameter to it so it can download data to `deps/yarn` upon request

Reply:
Ok, fixed